### PR TITLE
Fix ScrollID used for retrieval of APM Agents configuration

### PR DESCRIPTION
**Pull request by up2neck**
- Fixed broken APM Agents configuration cache

## Motivation/summary

APM Agents configuration cache appears to be broken, resulting frequent invalid requests made by APM Server to Elasticsearch cluster.

## Checklist

- [x] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/main/CHANGELOG.asciidoc)
- [x] Documentation has been updated

For functional changes, consider:
- Is it observable through the addition of either **logging** or **metrics**?
- Is its use being published in **telemetry** to enable product improvement?
- Have system tests been added to avoid regression?

## How to test these changes

1. Deploy Elastic Stack: Elasticsearch, Kibana, APM Server.
2. Configure self-instrumentation for APM Server
3. Open Kibana/APM
4. Check APM Server has no failed operations of type "Elasticsearch: POST .apm-agent-configuration/_search" for Elasticsearch dependency

## Related issues

closes [#13957](https://github.com/elastic/apm-server/issues/13957)

### CODE REVIEW

File: changelogs/head.asciidoc [hunk 1]
```diff
@@ -7,6 +7,7 @@ https://github.com/elastic/apm-server/compare/8.15\...main[View commits]
 ==== Bug fixes
 
 - Track all bulk request response status codes {pull}13574[13574]
+- Fix APM Agents configuration cache renewal {pull}13958[13958]
```
Comment (carsonip) [inline changelogs/head.asciidoc:10]: It is fine to remove this line, as this bugfix will likely be backported to 8.15.
```suggestion
```

**Review by carsonip — COMMENTED** [review]
Thanks for the PR. Nice catch!

Ideally this PR also updates the test to assert scroll id payload in `elasticsearch_test.go`, but it may not be trivial. I can do it in a follow up PR.

**Comment by up2neck** [comment]
@carsonip 
Alongside this, what do you think about adding check of "scroll_id" is not empty after cache renewal?

**Comment by carsonip** [comment]
> Alongside this, what do you think about adding check of "scroll_id" is not empty after cache renewal?

It doesn't hurt, but personally I do not think there is a _need_ in handler to validate anything returned by Elasticsearch. We rely on Elasticsearch for a correct scroll ID, as well as correct search results. What's needed here is a better test, which I'm working on.

Either way the error will be surfaced, just more explicitly if an additional check is in place, or a little more obscurely in a form of HTTP 400 like this case.

**Comment by up2neck** [comment]
> > Alongside this, what do you think about adding check of "scroll_id" is not empty after cache renewal?
> 
> It doesn't hurt, but personally I do not think there is a _need_ in handler to validate anything returned by Elasticsearch. We rely on Elasticsearch for a correct scroll ID, as well as correct search results. What's needed here is a better test, which I'm working on.
> 
> Either way the error will be surfaced, just more explicitly if an additional check is in place, or a little more obscurely in a form of HTTP 400 like this case.

But original code contains HTTP validation, but only for the first search request: 
https://github.com/elastic/apm-server/blob/a5aaf076514d6bd849eaf2bd8baed7cc5df5e845/internal/agentcfg/elasticsearch.go#L266

Probably we can modify it to handle both of the requests:
```
func (f *ElasticsearchFetcher) singlePageRefresh(ctx context.Context, scrollID string) (cacheResult, error) {
	var result cacheResult
	var err error
	var resp *esapi.Response

	if scrollID == "" {
		resp, err = esapi.SearchRequest{
			Index:  []string{ElasticsearchIndexName},
			Size:   &f.searchSize,
			Scroll: f.cacheDuration,
		}.Do(ctx, f.client)
	} else {
		resp, err = esapi.ScrollRequest{
			ScrollID: scrollID,
			Scroll:   f.cacheDuration,
		}.Do(ctx, f.client)
	}
	if err != nil {
		return result, err
	}
	defer resp.Body.Close()

	if resp.StatusCode >= http.StatusBadRequest {
		// Elasticsearch returns 401 on unauthorized requests and 403 on insufficient permission
		if resp.StatusCode == http.StatusUnauthorized || resp.StatusCode == http.StatusForbidden {
			f.invalidESCfg.Store(true)
		}
		bodyBytes, err := io.ReadAll(resp.Body)
		if err == nil {
			f.logger.Debugf("refresh cache elasticsearch returned status %d: %s", resp.StatusCode, string(bodyBytes))
		}
		return result, fmt.Errorf("refresh cache elasticsearch returned status %d", resp.StatusCode)
	}
	return result, json.NewDecoder(resp.Body).Decode(&result)
}
```

**Comment by carsonip** [comment]
> But original code contains HTTP validation, but only for the first search request

Correct. There is no handling for bad HTTP response codes in scroll requests. In retrospect, there should be. Thanks for adding it to the PR.

However, that increases the scope of this PR a little bit, and it will need to be tested on our end to avoid introducing regression (I don't expect any).

**Review by carsonip — DISMISSED** [review]
lgtm, thanks for the PR. I can confirm this fixes the scrolling bug by my follow-up PR to add tests: https://github.com/elastic/apm-server/pull/13959 . The refactoring to handle http errors should be fine, as expired scroll requests returns 404.

**Comment by carsonip** [comment]
run docs-build

**Comment by up2neck** [comment]
> > But original code contains HTTP validation, but only for the first search request
> 
> Correct. There is no handling for bad HTTP response codes in scroll requests. In retrospect, there should be. Thanks for adding it to the PR.
> 
> However, that increases the scope of this PR a little bit, and it will need to be tested on our end to avoid introducing regression (I don't expect any).

Made it

**Review by carsonip — APPROVED** [review]

**Comment by carsonip** [comment]
run docs-build

**Comment by up2neck** [comment]
@carsonip unfortunately, Github lost my commit signature after rebase in web UI. Have restored it.

**Review by carsonip — APPROVED** [review]

**Comment by carsonip** [comment]
run docs-build

**Comment by carsonip** [comment]
run docs-build

**Pull request merged**